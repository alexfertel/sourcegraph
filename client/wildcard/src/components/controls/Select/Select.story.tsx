import { Meta } from '@storybook/react'
import React, { useCallback } from 'react'

import { BrandedStory } from '@sourcegraph/branded/src/components/BrandedStory'
import webStyles from '@sourcegraph/web/src/SourcegraphWebApp.scss'

import { Select } from './Select'

const Story: Meta = {
    title: 'wildcard/Select',

    decorators: [
        story => (
            <BrandedStory styles={webStyles}>{() => <div className="container mt-3">{story()}</div>}</BrandedStory>
        ),
    ],

    parameters: {
        component: Select,
        design: {
            type: 'figma',
            name: 'Figma',
            url:
                'https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/Design-Refresh-Systemization-source-of-truth?node-id=908%3A1353',
        },
    },
}

// eslint-disable-next-line import/no-default-export
export default Story

export const Simple = () => {
    const [selected, setSelected] = React.useState('')

    const handleChange = useCallback<React.ChangeEventHandler<HTMLSelectElement>>(event => {
        setSelected(event.target.value)
    }, [])

    return (
        <Select label="What is your favorite fruit?" message="Hello world" value={selected} onChange={handleChange}>
            <option value="">Select a value</option>
            <option value="apples">Apples</option>
            <option value="bananas">Bananas</option>
            <option value="oranges">Oranges</option>
        </Select>
    )
}
