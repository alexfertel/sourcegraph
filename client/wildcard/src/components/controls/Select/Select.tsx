import classNames from 'classnames'
import React from 'react'

export interface SelectProps
    extends React.SelectHTMLAttributes<HTMLSelectElement>,
        React.RefAttributes<HTMLSelectElement> {
    className?: string
    isValid?: boolean
    label: React.ReactNode
    message?: React.ReactNode
}

export const Select: React.FunctionComponent<SelectProps> = React.forwardRef(
    ({ children, className, label, message, isValid, ...selectProps }, reference) => (
        <div className="form-check">
            <label className="form-check-label">
                <select ref={reference} className={classNames('form-control', className)} {...selectProps}>
                    {children}
                </select>
                {label}
            </label>
            {message && <small className={'field-message'}>{message}</small>}
        </div>
    )
)
