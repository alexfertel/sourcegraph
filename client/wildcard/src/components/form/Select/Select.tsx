import classNames from 'classnames'
import React from 'react'

import { FormFieldMessage } from '../internal/FormFieldMessage'

export interface SelectProps
    extends React.SelectHTMLAttributes<HTMLSelectElement>,
        React.RefAttributes<HTMLSelectElement> {
    className?: string
    /**
     * Used to control the styling of the <select> and surrounding elements.
     * Set this value to `false` to show invalid styling.
     * Set this value to `true` to show valid styling.
     */
    isValid?: boolean
    /**
     * Descriptive text rendered within a <label> element.
     */
    label: React.ReactNode
    /**
     * Optional message to display below the <select>.
     * This should typically be used to display additional information (perhap error/success states) to the user.
     * It will be styled differently if `isValid` is truthy or falsy.
     */
    message?: React.ReactNode
    /**
     * Use the Bootstrap custom <select> styles
     */
    isCustomStyle?: boolean
}

export const getSelectStyles = (isCustomStyle?: boolean): string => {
    if (isCustomStyle) {
        return 'custom-select'
    }

    return 'form-control'
}

export const Select: React.FunctionComponent<SelectProps> = React.forwardRef(
    ({ children, className, label, message, isValid, isCustomStyle, ...selectProps }, reference) => (
        <div className="form-group">
            <label>
                <select
                    ref={reference}
                    className={classNames(getSelectStyles(isCustomStyle), className)}
                    {...selectProps}
                >
                    {children}
                </select>
                {label}
            </label>
            {message && <FormFieldMessage isValid={isValid}>{message}</FormFieldMessage>}
        </div>
    )
)
