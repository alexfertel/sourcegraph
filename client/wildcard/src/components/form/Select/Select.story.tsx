import { Meta } from '@storybook/react'
import React, { useCallback } from 'react'

import { BrandedStory } from '@sourcegraph/branded/src/components/BrandedStory'
import webStyles from '@sourcegraph/web/src/SourcegraphWebApp.scss'

import { Select, SelectProps } from './Select'

const Story: Meta = {
    title: 'wildcard/Select',

    decorators: [
        story => (
            <BrandedStory styles={webStyles}>{() => <div className="container mt-3">{story()}</div>}</BrandedStory>
        ),
    ],

    parameters: {
        component: Select,
        design: {
            type: 'figma',
            name: 'Figma',
            url:
                'https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/Design-Refresh-Systemization-source-of-truth?node-id=908%3A1353',
        },
    },
}

// eslint-disable-next-line import/no-default-export
export default Story

const BaseSelect = (props: Partial<SelectProps>) => {
    const [selected, setSelected] = React.useState('')

    const handleChange = useCallback<React.ChangeEventHandler<HTMLSelectElement>>(event => {
        setSelected(event.target.value)
    }, [])

    return (
        <Select label="What is your favorite fruit?" value={selected} onChange={handleChange} {...props}>
            <option value="">Favorite fruit</option>
            <option value="apples">Apples</option>
            <option value="bananas">Bananas</option>
            <option value="oranges">Oranges</option>
        </Select>
    )
}

const SelectVariants = ({ isCustomStyle }: Pick<SelectProps, 'isCustomStyle'>) => (
    <>
        <h3>Simple</h3>
        <BaseSelect isCustomStyle={isCustomStyle} />
        <BaseSelect isCustomStyle={isCustomStyle} isValid={false} />
        <BaseSelect isCustomStyle={isCustomStyle} isValid={true} />

        <h3>With message</h3>
        <BaseSelect isCustomStyle={isCustomStyle} message="I am a message" />
        <BaseSelect isCustomStyle={isCustomStyle} message="I am a message" isValid={false} />
        <BaseSelect isCustomStyle={isCustomStyle} message="I am a message" isValid={true} />
    </>
)

export const SelectExamples: React.FunctionComponent = () => (
    <>
        <h1>Select</h1>
        <h2>Native select</h2>
        <SelectVariants />
        <h2>Custom select</h2>
        <SelectVariants isCustomStyle={true} />
    </>
)
