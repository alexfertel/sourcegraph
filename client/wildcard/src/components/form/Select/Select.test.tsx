import { render } from '@testing-library/react'
import React from 'react'

import { Select } from './Select'

describe('Select', () => {
    it('renders correctly', () => {
        const { container } = render(
            <Select label="What is your favorite fruit?" message="Hello world">
                <option value="">Select a value</option>
                <option value="apples">Apples</option>
                <option value="bananas">Bananas</option>
                <option value="oranges">Oranges</option>
            </Select>
        )

        expect(container.firstChild).toMatchInlineSnapshot(`
            <div
              class="form-check"
            >
              <label
                class="form-check-label"
              >
                <select
                  class="form-control"
                >
                  <option
                    value=""
                  >
                    Select a value
                  </option>
                  <option
                    value="apples"
                  >
                    Apples
                  </option>
                  <option
                    value="bananas"
                  >
                    Bananas
                  </option>
                  <option
                    value="oranges"
                  >
                    Oranges
                  </option>
                </select>
                What is your favorite fruit?
              </label>
              <small
                class="field-message"
              >
                Hello world
              </small>
            </div>
        `)
    })
})
