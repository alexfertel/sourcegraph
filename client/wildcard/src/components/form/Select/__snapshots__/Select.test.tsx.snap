// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select custom select renders correctly 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="custom-select"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
</div>
`;

exports[`Select custom select renders with correct styles when invalid 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="custom-select"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
  <small
    class="invalid-feedback"
  >
    Additional message
  </small>
</div>
`;

exports[`Select custom select renders with correct styles when valid 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="custom-select"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
  <small
    class="valid-feedback"
  >
    Additional message
  </small>
</div>
`;

exports[`Select custom select renders with message correctly 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="custom-select"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
  <small
    class="field-message"
  >
    Additional message
  </small>
</div>
`;

exports[`Select native select renders correctly 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="form-control"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
</div>
`;

exports[`Select native select renders with correct styles when invalid 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="form-control"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
  <small
    class="invalid-feedback"
  >
    Additional message
  </small>
</div>
`;

exports[`Select native select renders with correct styles when valid 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="form-control"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
  <small
    class="valid-feedback"
  >
    Additional message
  </small>
</div>
`;

exports[`Select native select renders with message correctly 1`] = `
<div
  class="form-group"
>
  <label>
    <select
      class="form-control"
    >
      <option
        value=""
      >
        Select a value
      </option>
      <option
        value="apples"
      >
        Apples
      </option>
      <option
        value="bananas"
      >
        Bananas
      </option>
      <option
        value="oranges"
      >
        Oranges
      </option>
    </select>
    What is your favorite fruit?
  </label>
  <small
    class="field-message"
  >
    Additional message
  </small>
</div>
`;
